File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/conformance/policy_runner.rb
# frozen_string_literal: true

require "json"
require_relative "result"
require_relative "../policy"

module IAIso
  module Conformance
    module PolicyRunner
      TOLERANCE = 1e-9

      module_function

      def run(spec_root)
        path = File.join(spec_root, "policy", "vectors.json")
        doc = ::JSON.parse(File.read(path))
        out = []
        (doc["valid"] || []).each { |v| out << run_valid(v) }
        (doc["invalid"] || []).each { |v| out << run_invalid(v) }
        out
      end

      def run_valid(v)
        name = "valid/#{v["name"]}"
        p = IAIso::Policy::Loader.build(v["document"])
        if v["expected_pressure"].is_a?(Hash)
          err = check_pressure(p, v["expected_pressure"])
          return VectorResult.fail("policy", name, err) if err
        end
        if v["expected_consent"].is_a?(Hash)
          err = check_consent(p, v["expected_consent"])
          return VectorResult.fail("policy", name, err) if err
        end
        if v["expected_metadata"].is_a?(Hash)
          if v["expected_metadata"].size != p.metadata.size
            return VectorResult.fail("policy", name, "metadata size: got #{p.metadata.size}, want #{v["expected_metadata"].size}")
          end
        end
        VectorResult.pass("policy", name)
      rescue => e
        VectorResult.fail("policy", name, "build failed: #{e.message}")
      end

      def check_pressure(p, ep)
        checks = {
          "token_coefficient"      => p.pressure.token_coefficient,
          "tool_coefficient"       => p.pressure.tool_coefficient,
          "depth_coefficient"      => p.pressure.depth_coefficient,
          "dissipation_per_step"   => p.pressure.dissipation_per_step,
          "dissipation_per_second" => p.pressure.dissipation_per_second,
          "escalation_threshold"   => p.pressure.escalation_threshold,
          "release_threshold"      => p.pressure.release_threshold,
        }
        checks.each do |k, got|
          next unless ep.key?(k)
          want = ep[k].to_f
          if (got - want).abs > TOLERANCE
            return "#{k}: got #{got}, want #{want}"
          end
        end
        if ep.key?("post_release_lock")
          return "post_release_lock mismatch" if ep["post_release_lock"] != p.pressure.post_release_lock
        end
        nil
      end

      def check_consent(p, ec)
        if ec.key?("issuer")
          if ec["issuer"] != p.consent.issuer
            return "consent.issuer: got #{p.consent.issuer.inspect}, want #{ec["issuer"].inspect}"
          end
        end
        if ec.key?("default_ttl_seconds")
          want = ec["default_ttl_seconds"].to_f
          if (p.consent.default_ttl_seconds - want).abs > TOLERANCE
            return "default_ttl_seconds: got #{p.consent.default_ttl_seconds}, want #{want}"
          end
        end
        if ec.key?("required_scopes")
          if ec["required_scopes"].size != p.consent.required_scopes.size
            return "required_scopes length mismatch"
          end
        end
        if ec.key?("allowed_algorithms")
          if ec["allowed_algorithms"].size != p.consent.allowed_algorithms.size
            return "allowed_algorithms length mismatch"
          end
        end
        nil
      end

      def run_invalid(v)
        name = "invalid/#{v["name"]}"
        IAIso::Policy::Loader.build(v["document"])
        VectorResult.fail("policy", name, "expected error containing '#{v["expect_error_path"]}', got Ok")
      rescue IAIso::Policy::PolicyError => e
        return VectorResult.fail("policy", name, "expected error containing '#{v["expect_error_path"]}', got: #{e.message}") unless e.message.include?(v["expect_error_path"])
        VectorResult.pass("policy", name)
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoConformance/PolicyRunner.swift
import Foundation
import IAIsoPolicy

/// Runs the policy section of the conformance suite.
public enum PolicyRunner {
    public static let TOLERANCE: Double = 1e-9

    public static func run(specRoot: String) throws -> [VectorResult] {
        let path = "\(specRoot)/policy/vectors.json"
        let data = try Data(contentsOf: URL(fileURLWithPath: path))
        guard let doc = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return [VectorResult.fail("policy", "load", "could not parse vectors.json")]
        }
        var out: [VectorResult] = []
        for v in (doc["valid"] as? [[String: Any]] ?? []) {
            out.append(runValid(v))
        }
        for v in (doc["invalid"] as? [[String: Any]] ?? []) {
            out.append(runInvalid(v))
        }
        return out
    }

    private static func runValid(_ v: [String: Any]) -> VectorResult {
        let name = "valid/\((v["name"] as? String) ?? "?")"
        do {
            let p = try PolicyLoader.build(v["document"])
            if let ep = v["expected_pressure"] as? [String: Any] {
                if let err = checkPressure(p, ep) {
                    return VectorResult.fail("policy", name, err)
                }
            }
            if let ec = v["expected_consent"] as? [String: Any] {
                if let err = checkConsent(p, ec) {
                    return VectorResult.fail("policy", name, err)
                }
            }
            if let em = v["expected_metadata"] as? [String: Any] {
                if em.count != p.metadata.count {
                    return VectorResult.fail("policy", name,
                        "metadata size: got \(p.metadata.count), want \(em.count)")
                }
            }
            return VectorResult.pass("policy", name)
        } catch {
            return VectorResult.fail("policy", name, "build failed: \(error)")
        }
    }

    private static func checkPressure(_ p: Policy, _ ep: [String: Any]) -> String? {
        let pairs: [(String, Double)] = [
            ("token_coefficient", p.pressure.tokenCoefficient),
            ("tool_coefficient", p.pressure.toolCoefficient),
            ("depth_coefficient", p.pressure.depthCoefficient),
            ("dissipation_per_step", p.pressure.dissipationPerStep),
            ("dissipation_per_second", p.pressure.dissipationPerSecond),
            ("escalation_threshold", p.pressure.escalationThreshold),
            ("release_threshold", p.pressure.releaseThreshold),
        ]
        for (k, got) in pairs {
            if let want = num(ep[k]), abs(got - want) > TOLERANCE {
                return "\(k): got \(got), want \(want)"
            }
        }
        if let want = ep["post_release_lock"] as? Bool,
           want != p.pressure.postReleaseLock {
            return "post_release_lock mismatch"
        }
        return nil
    }

    private static func checkConsent(_ p: Policy, _ ec: [String: Any]) -> String? {
        if ec.keys.contains("issuer") {
            let want = ec["issuer"] as? String
            if want != p.consent.issuer {
                return "consent.issuer: got \(p.consent.issuer ?? "nil"), want \(want ?? "nil")"
            }
        }
        if let want = num(ec["default_ttl_seconds"]),
           abs(p.consent.defaultTTLSeconds - want) > TOLERANCE {
            return "default_ttl_seconds: got \(p.consent.defaultTTLSeconds), want \(want)"
        }
        if let want = ec["required_scopes"] as? [Any],
           want.count != p.consent.requiredScopes.count {
            return "required_scopes length mismatch"
        }
        if let want = ec["allowed_algorithms"] as? [Any],
           want.count != p.consent.allowedAlgorithms.count {
            return "allowed_algorithms length mismatch"
        }
        return nil
    }

    private static func runInvalid(_ v: [String: Any]) -> VectorResult {
        let name = "invalid/\((v["name"] as? String) ?? "?")"
        let expectPath = (v["expect_error_path"] as? String) ?? ""
        do {
            _ = try PolicyLoader.build(v["document"])
            return VectorResult.fail("policy", name,
                "expected error containing '\(expectPath)', got Ok")
        } catch let err as PolicyError {
            if err.message.contains(expectPath) {
                return VectorResult.pass("policy", name)
            }
            return VectorResult.fail("policy", name,
                "expected error containing '\(expectPath)', got: \(err.message)")
        } catch {
            return VectorResult.fail("policy", name, "unexpected error: \(error)")
        }
    }

    private static func num(_ v: Any?) -> Double? {
        if let n = v as? NSNumber {
            #if canImport(ObjectiveC)
            if String(cString: n.objCType) == "c" { return nil }
            #endif
            return n.doubleValue
        }
        if let d = v as? Double { return d }
        if let i = v as? Int { return Double(i) }
        if let i = v as? Int64 { return Double(i) }
        return nil
    }
}
